source/simulator/core.cc

```diff
@@ -769,23 +791,77 @@ namespace aspect
             }
       }
 
+    if (!boundary_composition_manager.allows_fixed_composition_on_outflow_boundaries())
+      restore_outflow_boundary_ids(boundary_id_offset);
+
+    if (parameters.include_melt_transport)
+      melt_handler->add_current_constraints (new_current_constraints);
+
     // let plugins add more constraints if they so choose, then close the
     // constraints object
-    signals.post_constraints_creation(*this, current_constraints);
+    signals.post_constraints_creation(*this, new_current_constraints);
+
+    new_current_constraints.close();
+
+    // Now check if the current_constraints we just computed changed from before. Do this before the melt handler
+    // adds constraints for the melt cells, because they are allowed to change (and often do) without us having
+    // to reinit the sparsity pattern.
+    // If the mesh got refined and the size of the linear system changed, the old and new constraint
+    // matrices will have different entries, and we can not easily compare them.
+    // However, in case of mesh refinement we have to rebuild the matrix anyway, so we can skip the
+    // step that checks if constraints have changed.
+    bool mesh_has_changed = (current_constraints.get_local_lines().size()
+                             != new_current_constraints.get_local_lines().size())
+                            ||
+                            (current_constraints.get_local_lines()
+                             != new_current_constraints.get_local_lines());
+
+    if (!mesh_has_changed)
+      {
+        bool constraints_changed = false;
+
+#if DEAL_II_VERSION_GTE(9,0,0)
+        for (auto &row: current_constraints.get_lines())
+          {
+            if (!new_current_constraints.is_constrained(row.index))
+              {
+                constraints_changed = true;
+                break;
+              }
+          }
+#else
+        for (const auto row: current_constraints.get_local_lines())
+          {
+            if (current_constraints.is_constrained(row)
+                != new_current_constraints.is_constrained(row))
+              {
+                constraints_changed = true;
+                break;
+              }
+          }
+#endif
+
+        const bool any_constraints_changed = Utilities::MPI::sum(constraints_changed ? 1 : 0,
+                                                                 mpi_communicator) > 0;
+        if (any_constraints_changed)
+          rebuild_sparsity_and_matrices = true;
+      }
 
+#if DEAL_II_VERSION_GTE(9,0,0)
+    current_constraints.copy_from(new_current_constraints);
+#else
+    current_constraints.clear ();
+    current_constraints.reinit (introspection.index_sets.system_relevant_set);
+    current_constraints.merge (new_current_constraints);
     current_constraints.close();
+#endif
 
     // let the melt handler add its constraints once before we solve the porosity system for the first time
```

Comment:
You moved or removed the following lines (because you do not need them any more?), please also remove the comment.